WIP: API defnition for k8s dynamic client

droot opened the pull request:
Sharing an early version of the `kubeclient` 
Its a high level abstraction build on top of `client-go` libs providing basic CRUD operations on K8s objects.

Liujingfang1 commented on pkg/kubeclient/client.go:
```diff
@@ -0,0 +1,89 @@
+package kubeclient
```
Have you considered any other name, such as `client`?

droot replied:
Done.

Liujingfang1 commented on pkg/kubeclient/client.go:
```diff
@@ -0,0 +1,89 @@
+// provided, it uses the configPath to initialize a client else it assumes to be
+// running within a Kubernetes cluster and tries to initialize with cluster
+// configuration.
+func FromConfigOrCluster(configPath string) (Interface, error) {
```
Is `FromConfigOrCluster` going to call `FromConfig` and `IsCluster`?

apelisse replied:
`Cluster` is confusing if you mean `InCluster`. Also the documentation reads `NewFrom...` but the method name is `From...`. TBH, I like the `NewFrom` form better. (also applies to the function below).

apelisse replied:
> and tries to initialize with **in**-cluster configuration

apelisse replied:
> If configPath is provided, it uses the configPath to initialize a client else it assumes

Why would I call this method if I don't provide a `configPath`, rather than calling the `InCluster` directly?

droot replied:
Now we have only one constructor which takes rest.Config as input. There are already helper methods in client-go which lets you create rest.Config.

droot commented:
@pwittrock couple of changes: 

- Updated the `List` method. Now it takes runtime.Object as input which is actually a List-type and added an example usage.
-  Removed the `Watch` method from the client `interface` because its usage is coupled with `Informers` which is an implementation detail of `controller-lib` and we can construct a watcher using dynamic-client. So as far as users are concerned, they don't need to know about the whole `watch` thing ever.

PTAL a look.

Liujingfang1 commented on pkg/kubeclient/client.go:
```diff
@@ -0,0 +1,89 @@
+
+// Delete deletes the given obj from Kubernetes cluster. obj must have TypeMeta and
+// ObjectMeta field populated.
+func (c *Client) Delete(ctx context.Context, obj runtime.Object) error {
```
This function is inconsistent with the interface, missing `opts *metav1.DeleteOptions`

droot replied:
This should be fixed now.

apelisse commented on pkg/kubeclient/client.go:
```diff
@@ -0,0 +1,89 @@
+	Create(ctx context.Context, obj runtime.Object) error
+
+	// Get retrieves given object from Kubernetes cluster
+	Get(ctx context.Context, obj runtime.Object) error
```
"retrieves given object". Could you be more specific? The hidden question here is, why do I need an object to get something?

droot replied:
One of the thing which is sort of hidden is that we will be using methods defined in runtime.Object to determine REST endpoint for the resource involved. We could take those as input explicitly also, but then there will be duplication and verbosity in the caller's code.

apelisse commented on pkg/kubeclient/client.go:
```diff
@@ -0,0 +1,89 @@
+	Update(ctx context.Context, obj runtime.Object) error
+
+	// Delete deletes given object from Kubernetes cluster
+	Delete(ctx context.Context, obj runtime.Object, opts *metav1.DeleteOptions) error
```
Same as get, do I need to get the object before I can delete? etc

droot replied:
Same comment as above for the `Get`.

apelisse commented on pkg/kubeclient/client.go:
```diff
@@ -0,0 +1,89 @@
+	Delete(ctx context.Context, obj runtime.Object, opts *metav1.DeleteOptions) error
+
+	// List retrieves list of object
+	List(ctx context.Context, obj runtime.Object, opts *metav1.ListOptions) error
```
I think that might be even more confusing :-)

apelisse commented on pkg/kubeclient/client.go:
```diff
@@ -0,0 +1,89 @@
+)
+
+// Interface defines capability of the Kubernetes Client.
+type Interface interface {
```
Don't we have an interface similar to that somewhere already? How is this interface better?

droot replied:
Not that I know of.

apelisse commented on pkg/kubeclient/client.go:
```diff
@@ -0,0 +1,89 @@
+
+// NewFromConfig returns a Kubernetes client given a path to the config file to
+// connect to Kubernetes cluster.
+func FromConfig(configPath string) (Interface, error) {
```
`New...`

droot replied:
Changed it completely.

apelisse commented on pkg/kubeclient/client.go:
```diff
@@ -0,0 +1,89 @@
+
+}
+
+// InCluster() uses default cluster configuration to initialize a Kubernetes
```
`NewFromInCluster`?

apelisse replied:
So you have 3 constructors, one that does from config, one that does from in-cluster, and one that does both. Can we reduce that number? The more methods you export, the more complicated the package.

droot replied:
Now we have only one constructor. PTAL.

apelisse commented on pkg/kubeclient/client.go:
```diff
@@ -0,0 +1,89 @@
+}
+
+// Client is a concrete implementation of Kubernetes Çlient interface.
+type Client struct {
```
Should this be exposed?

apelisse commented on pkg/kubeclient/client.go:
```diff
@@ -0,0 +1,89 @@
+// in the obj and list options. obj must have populated TypeMeta field. On a
+// successful call, Items field in the obj will be populated with the returned
+// contained.
+func (c *Client) List(ctx context.Context, namespace string, obj runtime.Object, opts *metav1.ListOptions) error {
```
Thank you Jingfang, this is also inconsistent with the interface. And looks like a mix of `obj` and specifying fields ;-)

apelisse commented on pkg/kubeclient/client.go:
```diff
@@ -0,0 +1,89 @@
+type Client struct {
+}
+
+// Create saves the object obj in to the Kubernetes cluster. obj must have
```
If you don't expose the client, then these comments belong to the interface, which I think would be better anyway since you're not really describing the specifics of that particular client, but the interface.

droot replied:
Totally makes sense. Done.

apelisse commented on pkg/kubeclient/client_test.go:
```diff
@@ -0,0 +1,106 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func ExampleClient_Create() {
```
Feel free to document your example by adding a comment befor e.

droot replied:
Collapsed all the examples in one example because there was lot of repetition. Also added comments.

apelisse commented on pkg/kubeclient/client_test.go:
```diff
@@ -0,0 +1,106 @@
+)
+
+func ExampleClient_Create() {
+	kc, err := kubeclient.FromConfigOrCluster("/home/dir/config/.kubeconfig")
```
does it have a default or should I always explicitly specify the location?

droot replied:
Now config is passed instead, so this is handled by the caller.

apelisse commented on pkg/kubeclient/client_test.go:
```diff
@@ -0,0 +1,106 @@
+	if err != nil {
+		// handle err
+	}
+}
```
While I like examples, I'm wondering if this is too many? The methods are pretty simple and straightforward ... WDYT?

apelisse replied:
Focusing on "how do these work together" should be good enough (package level example)

droot replied:
Totally agree. Collapsed all examples in to one pkg level example.

apelisse commented on pkg/kubeclient/doc.go:
```diff
@@ -0,0 +1,56 @@
+
+
+List Kubernetes Objects:
+	ctx := context.Background()
```
should these be examples?

droot replied:
I have pkg level example and duplicated in doc.go as well. I will take it out from doc.go after seeing how godocs looks for it.

droot commented:
Thanks for the comments @Liujingfang1 @apelisse , I am out of most of the day, will be addressing these tomorrow.

droot commented on pkg/client/client.go:
```diff
@@ -0,0 +1,72 @@
+)
+
+// ObjectKey identifies a Kubernetes Object.
+type ObjectKey types.NamespacedName
```
I am still thinking if it is an overkill ? because passing in `namespace` and `name` as is is less verbose.

apelisse replied:
Yeah, I've had a very long discussion with Phil about that, still believe that `types.NamespacedName` is more explicit than `ObjectKey`.

apelisse replied:
what's wrong with taking `namespace, name string`?

droot replied:
Nothing wrong with taking `namespace` and `name` input, but I want to look at it from `ctrl` pkg's point of view as well because this pkg is going to be consumed by the `ctrl` pkg and would like it to be consistent.

DirectXMan12 replied:
Plus, taking `NamespacedName` makes it a *bit* harder to screw up the ordering of the two, especially if you have a linter that warns about initializing structs without field names.

DirectXMan12 replied:
perhaps it should be an alias though `type ObjectKey = types.NamespacedName`

droot commented:
cc @fanzhangio @hasbro17

apelisse commented on pkg/client/client.go:
```diff
@@ -0,0 +1,72 @@
+type ObjectKey types.NamespacedName
+
+// Client uses methods defined in runtime.Object to determine the REST endpoint
+// for Kubernetes resources.
```
Why is this comment dangling? Does it end-up in the godoc? What's "Client"?

droot replied:
Aah... I forgot to take it out. This is out of place. I need to put it somewhere but in context.

apelisse commented on pkg/client/client.go:
```diff
@@ -0,0 +1,72 @@
+	// List retrieves list of objects for a given namespace and list options. On a
+	// successful call, Items field in the list will be populated with the
+	// result returned from the server.
+	List(ctx context.Context, namespace string, opts *metav1.ListOptions, list runtime.Object) error
```
What is the reason why you're modifying the list in-place? instead of returning it?

droot replied:
Same reasoning as above. The hidden element here is that `obj` gives us `gvk` input which is causing this issue.

apelisse replied:
Can you create a GVKN and have a method to parse an object into a GVKN?
At this point, you barely have two methods that are consistent in usage, which is very confusing. I've always been confused in the past by libraries that took empty object to guess what to do.

DirectXMan12 replied:
One of the aims was to not have to specify a GVK when we didn't need to/it was redundant, and to not require type assertions.  The user already knows what kind of object they're using, it seems silly to have them write that twice.

The actual output code is fairly readable, IMO:

```go
var pods kapi.PodList
if err := client.List(context.TODO(), "somenamespace", nil, &pods); err != nil {
   return err
}
```

as opposed to with GVK:

```
obj, err := client.List(context.TODO(), "foo", nil, schema.GroupVersionKind{Version: "v1", Kind: "Pod")
if err != nil {
   return err
}
podList := obj.(*kapi.PodList) // hope that this is correct and that you didn't make a typo
```

apelisse commented on pkg/client/client.go:
```diff
@@ -0,0 +1,72 @@
+	List(ctx context.Context, namespace string, opts *metav1.ListOptions, list runtime.Object) error
+}
+
+// NewForConfig returns a Kubernetes client implementation given a
```
from?

droot replied:
For or From both works (I have seen both being used in APIMachinery). I don't have any strong preference. which one do you prefer ?

apelisse commented on pkg/client/client.go:
```diff
@@ -0,0 +1,72 @@
+	Update(ctx context.Context, obj runtime.Object) error
+
+	// Delete deletes the given obj from Kubernetes cluster.
+	Delete(ctx context.Context, obj runtime.Object) error
```
That's sort of inconsistent with `Get`?

droot replied:
I am sort of struggling with it. Taking just `key` is not enough, because we need gvk info for deleting the object and `obj` helps in giving us that info. And taking both `key` and `obj` is worse than taking just the `obj`

apelisse commented on pkg/client/doc.go:
```diff
@@ -0,0 +1,62 @@
+
+Create, Get, Update, Delete and List Kubernetes Objects:
+
+	// use helper methods to build config for client.
```
Isn't this 100% redundant with your example?

droot replied:
Yes. I wanted to see in godoc if the `Example` shows up in overview or not. Will take it out if it does.

droot commented on pkg/client/doc.go:
```diff
@@ -0,0 +1,62 @@
+		// print p
+	}
+*/
+package kubeclient
```
this needs to be `client`

hasbro17 commented on pkg/client/client.go:
```diff
@@ -0,0 +1,330 @@
+
+	// TODO(droot): figure out why GetObjectKind() is not working for an empty
+	// object ? Shouldn't it use scheme internally. If not then, there isn't
+	// much benefit in taking the runtime.Object
```
I should have pointed this out earlier. We had to enforce specifying the TypeMeta for the object in our implementation:
```Go
pod := &v1.Pod{
	TypeMeta: metav1.TypeMeta{
		Kind:       "Pod",
		APIVersion: "v1",
	},
}
podKey := client.ObjectKey{Namespace: "ns-1", Name: "pod-1"}
err = kc.Get(ctx, podKey, pod)
if err != nil {
	// handle err
}
```
That makes it more cumbersome to use though. Using the scheme to get the GVK is fine if it lets us pass in an empty object.

However we still need to pass in the runtime.Object to unmarshal the unstructured object into it. I don't think we can do away with that.

droot replied:
> That makes it more cumbersome to use though. Using the scheme to get the GVK is fine if it lets us pass in an empty object.

Yes, thats the idea. Using scheme, we are able to work with empty object and caller code looks very simple. Yes, we can't do away with the runtime.obj and the good thing all generated types support runtime.Obj interface, so we are good.

hasbro17 commented on pkg/client/client.go:
```diff
@@ -0,0 +1,330 @@
+	c := &client{
+		conf:       config,
+		dc:         dc,
+		scheme:     scheme.Scheme,
```
The standard client-go scheme is used over here.
Is it assumed that any custom types defined elsewhere need to add their types to the client-go scheme before the dynamic-client can correctly map the custom type to its GVK?
https://github.com/kubernetes-sigs/kubebuilder/blob/master/samples/controller/pkg/apis/samplecontroller/v1alpha1/register.go#L42

```Go
import (
	samplecontrollerv1alpha1 "github.com/kubernetes-sigs/kubebuilder/samples/controller/pkg/apis/samplecontroller/v1alpha1"
	"k8s.io/client-go/kubernetes/scheme"
)

func init() {
	samplecontrollerv1alpha1.AddToScheme(scheme.Scheme)
}

```

droot replied:
> Is it assumed that any custom types defined elsewhere need to add their types to the client-go scheme before the dynamic-client can correctly map the custom type to its GVK?

No, would like to avoid that. I am thinking more along the lines of providing a hook to register the custom-type to the scheme of the client. will be trying that out today.

mengqiy commented on pkg/client/client.go:
```diff
@@ -0,0 +1,330 @@
+		panic(err)
+	}
+	var u unstructured.Unstructured
+	if err := json.Unmarshal(b, &u.Object); err != nil {
```
You can use `u.UnmarshalJSON(b)` here.

droot replied:
+1

droot commented on pkg/client/client.go:
```diff
@@ -0,0 +1,330 @@
+// NewForConfig returns a Kubernetes client implementation given a
+// configuration.
+func NewForConfig(config *rest.Config) (Interface, error) {
+	if config == nil {
```
That introduces ambiguity in the interface, so would like to avoid that. client-go has many helpers functions to construct this config, so it should be easy. Also, godoc for this API documents one of the ways to create config..

droot commented on pkg/client/client.go:
```diff
@@ -0,0 +1,330 @@
+	return c, nil
+}
+
+// Client is a concrete implementation of Kubernetes Çlient interface.
```
I have that at the bottom. will move it here.

Liujingfang1 commented on pkg/client/client.go:
```diff
@@ -0,0 +1,330 @@
+
+// GetNameAndNamespace extracts the name and namespace from the given runtime.Object
+// and returns a error if any of those is missing.
+func namespaceAndName(object runtime.Object) (string, string, error) {
```
The comment is inconsistent with your function name.

Liujingfang1 commented on pkg/client/client.go:
```diff
@@ -0,0 +1,330 @@
+	return nil
+}
+
+// apiResource consults the REST mapper to translate an <apiVersion, kind, namespace> tuple to a schema.GroupVersionResource struct.
```
This comment looks confusing to me. The function accepts `gvk`, I don't see where `namesapce` comes from.

Liujingfang1 commented on pkg/client/client.go:
```diff
@@ -0,0 +1,330 @@
+}
+
+// UnstructuredFromRuntimeObject converts a runtime object to an unstructured
+func unstructuredFromRuntimeObject(ro runtime.Object) *unstructured.Unstructured {
```
Maybe change this function to `unstructuredFromObject` to be consistent with `unstructuredToObject`

droot commented:
Thanks for the comments everyone. @DirectXMan12 has worked on another implementation in controller-platform branch (https://github.com/kubernetes-sigs/kubebuilder/blob/controller-platform/pkg/client/client.go)  which is much cleaner and uses the right building-blocks from api-machinery under the hood. So for now, will pause on this PR.

droot commented:
closing this because we have an implementation in https://github.com/kubernetes-sigs/controller-runtime  repo.
